server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -859,21 +876,59 @@ private static Repository createRepository(
      * @return the started repository
      * @throws RepositoryException if repository type is not registered
      */
+    @FixForMultiProject(description = "resolve the actual ProjectId")
+    @Deprecated(forRemoval = true)
     public Repository createRepository(RepositoryMetadata repositoryMetadata) {
-        return createRepository(repositoryMetadata, typesRegistry, RepositoriesService::throwRepositoryTypeDoesNotExists);
+        return createRepository(ProjectId.DEFAULT, repositoryMetadata);
     }
 
-    private static Repository throwRepositoryTypeDoesNotExists(RepositoryMetadata repositoryMetadata) {
-        throw new RepositoryException(repositoryMetadata.name(), "repository type [" + repositoryMetadata.type() + "] does not exist");
+    /**
+     * Creates a repository holder.
+     *
+     * <p>WARNING: This method is intended for expert only usage mainly in plugins/modules. Please take note of the following:</p>
+     *
+     * <ul>
+     *     <li>This method does not register the repository (e.g., in the cluster state).</li>
+     *     <li>This method starts the repository. The repository should be closed after use.</li>
+     *     <li>The repository metadata should be associated to an already registered non-internal repository type and factory pair.</li>
+     * </ul>
+     *
+     * @param projectId the project that the repository is associated with
+     * @param repositoryMetadata the repository metadata
+     * @return the started repository
+     * @throws RepositoryException if repository type is not registered
+     */
+    public Repository createRepository(ProjectId projectId, RepositoryMetadata repositoryMetadata) {
+        return createRepository(
+            Objects.requireNonNull(projectId),
+            repositoryMetadata,
+            typesRegistry,
+            RepositoriesService::throwRepositoryTypeDoesNotExists
+        );
+    }
+
+    /**
+     * Similar to {@link #createRepository(ProjectId, RepositoryMetadata)}, but repository is not associated with a project, i.e. the
+     * repository is at the cluster level.
+     */
+    public Repository createNonProjectRepository(RepositoryMetadata repositoryMetadata) {
+        return createRepository(null, repositoryMetadata, typesRegistry, RepositoriesService::throwRepositoryTypeDoesNotExists);
```

Comment:
Minor comment really, but we could assert that we are running stateless here - I would expect us never to use this beyond stateless and at least an assert could help avoid someone using it by mistake.

```
assert DiscoveryNode.isStateless(clusterService.getSettings()) : "outside stateless only project level repositories are allowed: " + repositoryMetadata;
```